File: 101_SymmetricTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isSymmetric(TreeNode root) {
        if(root == null) {
        	return true;
        }
        else {
        	return checkLR(root.left, root.right);
        }
    }
    public boolean checkLR(TreeNode l, TreeNode r) {
		if(l == null && r == null) {
			return true;
		}
		if(l != null && r != null && l.val == r.val) {
			return checkLR(l.left, r.right) && checkLR(l.right, r.left);
		}
		return false;
	}
}

File: c++/101_SymmetricTree.cpp
// sol1
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        return helper(root, root);
    }
    
    bool helper(TreeNode* p, TreeNode* q) {
        if (p == nullptr && q == nullptr) {
            return true;
        }
        if (p == nullptr || q == nullptr) {
            return false;
        }
        return (p->val == q->val) && helper(p->left, q->right) && helper(p->right, q->left);
    }
};
// sol2
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        return helper(root, root);
    }
    
    bool helper(TreeNode* p, TreeNode* q) {
        queue<TreeNode*> que;
        que.push(p);
        que.push(q);
        while (!que.empty()) {
            p = que.front();
            que.pop();
            q = que.front();
            que.pop();
            if (p == nullptr && q == nullptr) {
                continue;
            }
            if ((p == nullptr || q == nullptr) || (p->val != q->val)) {
                return false;
            }
            
            que.push(p->left);
            que.push(q->right);
            
            que.push(p->right);
            que.push(q->left);
        }
        return true;
    }
};
